include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,811 @@
+/* Copyright 2013-2019 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
+ *                     Klaus Steiniger, Felix Schmitt, Benjamin Worpitz,
+ *                     Juncheng E, Pawel Ordyna
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+#include "picongpu/particles/traits/SpeciesEligibleForSolver.hpp"
+#include "picongpu/plugins/ISimulationPlugin.hpp"
+#include "picongpu/plugins/common/stringHelpers.hpp"
+
+#include "picongpu/fields/FieldTmp.hpp"
+#include "picongpu/param/saxs.param"
+#include "picongpu/plugins/saxs/beam/SaxsProbingBeam.hpp"
+#include "picongpu/plugins/saxs/Saxs.kernel"
+#include "picongpu/plugins/saxs/SaxsWriter.hpp"
+#include "picongpu/plugins/saxs/saxsUtilities.hpp"
+#include "picongpu/plugins/saxs/GetScatteringVector.hpp"
+#include "picongpu/plugins/saxs/traits/DetermineFormFactorSolver.hpp"
+#include "picongpu/particles/particleToGrid/derivedAttributes/Density.def"
+
+#include <pmacc/dataManagement/DataConnector.hpp>
+#include <pmacc/dimensions/DataSpaceOperations.hpp>
+#include <pmacc/mappings/kernel/AreaMapping.hpp>
+#include <pmacc/memory/MakeUnique.hpp>
+#include <pmacc/mpi/MPIReduce.hpp>
+#include <pmacc/mpi/reduceMethods/Reduce.hpp>
+#include <pmacc/nvidia/functors/Add.hpp>
+#include <pmacc/traits/GetNumWorkers.hpp>
+#include <pmacc/traits/HasFlag.hpp>
+#include <pmacc/assert.hpp>
+
+#include <boost/filesystem.hpp>
+#include <boost/mpl/bool.hpp>
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace picongpu
+{
+namespace plugins
+{
+namespace saxs
+{
+
+    using namespace pmacc;
+    using namespace picongpu::plugins::saxs;
+    using namespace picongpu::SI;
```

Comment:
I think you are right since this is going to influence other files as well.  Do you think it is a good practice to do it for the current namespace only? I mean, should I let  `using namespace picongpu::plugins::saxs;` stay?

Reply:
Ah, I actually missed the fact that `using namespace picongpu::plugins::saxs;` is actually the same namespace. In this case this line here does nothing (you are already in this namespace and so all its names are already here) and so can be safely removed. About the other two: what they do is indeed affect all files including this one. However, this is somewhat limited by the fact that you do it only for your saxs namespace, so it's not that disruptive. I guess since the implementation is quite large it's probably not worth it to find and explicitly using-declare only the names actually used, and so the other two  `using namespace` are okay here (and, as I already mentioned, we do it in many other places anyway, although I try to avoid it in new code).